### DISCUSSION TITLE
Is there a way to do "best-effort" XMP metadata parsing? (to overcome invalid XML)

### DISCUSSION THREAD

**Discussion by ribtoks** [Exiv2/exiv2#3068] [Q&A] [answered]
There's a file that has a corrupted XMP packet, which can be read by exiftool, but exiv2 says

```
Error: XMP Toolkit error 201: Error in XMLValidator
Warning: Failed to decode XMP metadata.
```

I understand that it's broken and it's OK. But maybe there's a way to overcome this?

I'm using exiv2 as a C++ library.

**Comment by clanmills** [comment]
@ribtoks There are options to delete/print/insert raw XMP packets.

```
$ exiv2 -pX https://clanmills.com/Stonehenge.jpg | xmllint --pretty 1 -
<?xml version="1.0"?>
<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 4.4.0-Exiv2">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description xmlns:xmp="http://ns.adobe.com/xap/1.0/" xmlns:cm2e="http://clanmills.com/exiv2/" xmlns:dc="http://purl.org/dc/elements/1.1/" rdf:about="" xmp:Rating="0" xmp:ModifyDate="2015-07-16T20:25:28+01:00" cm2e:Father="Robin Mills">
      <cm2e:Family>
        <rdf:Bag/>
      </cm2e:Family>
      <dc:description>
        <rdf:Alt>
          <rdf:li xml:lang="x-default">Classic View</rdf:li>
        </rdf:Alt>
      </dc:description>
      <dc:Family>
        <rdf:Bag>
          <rdf:li>Robin</rdf:li>
        </rdf:Bag>
      </dc:Family>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
<?xpacket end="w"?>
```

Before you can read that XMP, could you figure out why the validator rejected it? Once you understand, it may be possible to modify the validator code to forgive and accept.

**Reply by clanmills** [reply]
Something similar was discussed last year: https://github.com/Exiv2/exiv2/discussions/3018

Using the output of the command: `$ exiv2 -pS https://clanmills.com/Stonehenge.jpg` you can see the location of the XMP _segment_ in the JPG and use `dd` to remove/rewrite the invalid XMP.

I discuss using `exiv2 -pS` and `dd` in my book: https://clanmills.com/exiv2/book/#3-1

**Reply by ribtoks** [reply]
Hi Robin. Hope you're doing well!

Thank you for the idea. I tried and found out that somebody "creatively" tried to write IPTC tags in XMP packet. That is still a valid XML, but tags are arbitrary.

The question is how to "ignore" that using exiv2?

```bash
~/Photos//invalid XMP » ~/Applications/exiv2/bin/exiv2 -pX test.jpg | xmllint --pretty 1 -

-:6: namespace error : Namespace prefix iptc on ObjectName is not defined
    <iptc:ObjectName>Redacted...</iptc:ObjectName>
                    ^
-:8: namespace error : Namespace prefix iptc on Caption-Abstract is not defined
    <iptc:Caption-Abstract>Redacted...
                          ^
-:10: namespace error : Namespace prefix iptc on Keywords is not defined
    <iptc:Keywords>Redacted...
                  ^
<?xml version="1.0"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description xmlns:dc="http://purl.org/dc/elements/1.1/" rdf:about="">
      <dc:title>Redacted...</dc:title>
      <iptc:ObjectName>Redacted...</iptc:ObjectName>
      <dc:description>Redacted...</dc:description>
      <iptc:Caption-Abstract>Redacted...</iptc:Caption-Abstract>
      <dc:subject>Redacted...</dc:subject>
      <iptc:Keywords>Redacted...</iptc:Keywords>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
```

**Reply by clanmills** [reply]
The person who embedded that code has seriously damaged that file.  The simplest fix is to totally remove the XMP with the option `$ exiv2 -dX pathname`

If you know how to repair the XMP, you can insert it.  I think the command is `$ cat repaired.xml | exiv2 -iX- pathname`

Two possible fixes for the XMP:
1.  Remove every: `<iptc:ABCD>....</iptc:ABCD>`
2. Add `xmlns:iptc=....` into the XML header `<x:xmpmeta...>
Option 2 might not work because the Adobe XMPsdk embedded in exiv2 is hostile to unexpected namespaces!

Anyway, you can extract the XMP with `exiv2 -pX pathname`, manually edit it, delete and insert your edited xmp.  You said that exiftool handles this file with grace, so inspecting the output of exiftool might give you other ideas.

From your comments, I believe you're using libexiv2 in an application, so you probably want a fix for this in the C++ code.  Exiv2 locates the XMP in the file and reads it into a memory buffer which it passes to the Adobe XMP.  You could add code to modify that buffer.  If you adopt strategy 1 Remove `<iptc:ABCD>....</iptc:ABCD>`, simply replace the string with white space.

In exiv2 0.27-maintenance the code is in jpgimage.cpp line 421:

```
            else if (   !foundXmpData
                     && marker == app1_
                     && size >= 31  // prevent out-of-bounds read in memcmp on next line
                     && memcmp(buf.pData_ + 2, xmpId_, 29) == 0) {
                xmpPacket_.assign(reinterpret_cast<char*>(buf.pData_ + 31), size - 31);
                -- REMOVE IPTC XML HERE --
                if (xmpPacket_.size() > 0 && XmpParser::decode(xmpData_, xmpPacket_)) {
#ifndef SUPPRESS_WARNINGS
                    EXV_WARNING << "Failed to decode XMP metadata.\n";
#endif
                }
                --search;
                foundXmpData = true;
            }
```
Of course, the correct code location to edit the buffer is `XmpParser::decode()` which is called by every image format.

I realise that this is a challenging undertaking. However, that's the way it is. One additional remark: nobody can guarantee that Team Exiv2 will accept your patch into the code base.

**Reply by clanmills** [reply]
I'm up early this morning and had an idea:

**Fix 1 Works** 
Add a namespace definition for iptc.
```
557 rmills@rmillsm1:~/temp $ xmllint --pretty 1 foo.xml 
<?xml version="1.0"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/" xmlns:iptc="clanmills:ns:iptc">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description xmlns:dc="http://purl.org/dc/elements/1.1/" rdf:about="">
      <dc:title>Redacted...</dc:title>
      <iptc:ObjectName>Redacted...</iptc:ObjectName>
      <dc:description>Redacted...</dc:description>
      <iptc:Caption-Abstract>Redacted...</iptc:Caption-Abstract>
      <dc:subject>Redacted...</dc:subject>
      <iptc:Keywords>Redacted...</iptc:Keywords>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
558 rmills@rmillsm1:~/temp $ curl -O https://clanmills.com/Stonehenge.jpg
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 6599k  100 6599k    0     0  3459k      0  0:00:01  0:00:01 --:--:-- 3458k
$ exiv2 -dX Stonehenge.jpg 
$ cat foo.xml | exiv2 -iX- Stonehenge.jpg 
$ exiv2 -px Stonehenge.jpg 
Xmp.dc.title                                 LangAlt     1  lang="x-default" Redacted...
Xmp.dc.description                           LangAlt     1  lang="x-default" Redacted...
Xmp.dc.subject                               XmpBag      1  Redacted...
Xmp.iptc.ObjectName                          XmpText    11  Redacted...
Xmp.iptc.Caption-Abstract                    XmpText    11  Redacted...
Xmp.iptc.Keywords                            XmpText    11  Redacted...
562 rmills@rmillsm1:~/temp $ 
```

**Fix 2 Works** 
Blank out the iptc: entries in the XML:
```
$ bbedit "foo.xml"
$ cat foo.xml
<?xml version="1.0"?>
<x:xmpmeta xmlns:x="adobe:ns:meta/">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description xmlns:dc="http://purl.org/dc/elements/1.1/" rdf:about="">
      <dc:title>Redacted...</dc:title>
                                                       
      <dc:description>Redacted...</dc:description>
                                              
      <dc:subject>Redacted...</dc:subject>
                                                  
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
$ cat foo.xml | exiv2 -iX- Stonehenge.jpg 
$ exiv2 -px Stonehenge.jpg 
Xmp.dc.title                                 LangAlt     1  lang="x-default" Redacted...
Xmp.dc.description                           LangAlt     1  lang="x-default" Redacted...
Xmp.dc.subject                               XmpBag      1  Redacted...
$ 
```

Editing `xmpPacket` in C++ is a couple of lines of C++.   Be careful:

1.  adding `xmlns:iptc="clanmills:ns:iptc"` will expand xmpPacket
2.  don't use iptc:ns:clanmills.  I made that up for testing.  There might an existing suitable NS:
```
$ exiv2 -vV | grep -i iptc
xmlns=Iptc4xmpCore:http://iptc.org/std/Iptc4xmpCore/1.0/xmlns/
xmlns=Iptc4xmpExt:http://iptc.org/std/Iptc4xmpExt/2008-02-29/
xmlns=acdsee:http://ns.acdsee.com/iptc/1.0/
$ 
```

**Reply by ribtoks** [reply]
Hi Robin. Thank you for your ideas! 
Is there a way to "pre-add" iptc namespace for parsing (e.g. "register" it)? Remember, I'm doing it from C++ with exiv2 as a library. I can do these "push-ups" for this particular file in my command line, but the fix in question should be a generic one, that will be deployed in my software. That's why I framed the question in the beginning as "is it possible to ignore XML parsing errors" as, if it was, it would allow to parse well-known fields and ignore unknown ones.

**Reply by ribtoks** [reply]
an update regarding your "Fix 2" I cannot edit `xmpPacket` in C++ as parsing fails before that... Was there a way to extract the whole packet in C++ as a bytes buffer?

**Reply by clanmills** [reply, marked as answer]
@ribtoks  The XML is invalid, so every XML parser (including xmllint) will barf.  You can't forgive or trick it, you have to repair the XML code.

`readMetadata()` reads the raw XML from the file.  
```
            // Read the rest of the segment.
            DataBuf buf(size);
            if (size > 0) {
                readOrThrow(*io_, buf.pData_ + 2, size - 2, kerFailedToReadImageData);
                memcpy(buf.pData_, sizebuf, 2);
            }
```
At that moment, buf.pData_ has something like this:

```
1012 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ exiv2 -pS https://clanmills.com/Stonehenge.jpg
STRUCTURE OF JPEG FILE: https://clanmills.com/Stonehenge.jpg
 address | marker       |  length | data
       0 | 0xffd8 SOI  
       2 | 0xffe1 APP1  |   15272 | Exif..II*......................
   15276 | 0xffe1 APP1  |    2786 | http://ns.adobe.com/xap/1.0/.<?x
                             ----   --------------------------------------->
   18064 | 0xffed APP13 |      96 | Photoshop 3.0.8BIM.......'.....
   18162 | 0xffe2 APP2  |    4094 | MPF.II*...............0100.....
   22258 | 0xffdb DQT   |     132 
   22392 | 0xffc0 SOF0  |      17 
   22411 | 0xffc4 DHT   |     418 
   22831 | 0xffda SOS  
1013 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ 
```
The little dance in the code with 29 and 31 are there to ignore the 29 bytes of "http://ns.adob......" PLUS the 2 bytes containing the `unsigned short` segment length (in this case 2786).

Andreas wrote that code about 20 years ago, so I don't know exactly why we have `xmpPacket_` AND `xmpData_`. A little digging with the debugger will tell you.

You'll need to add code between the `readOrThrow()` and sending the data to Adobe.    Something like this:
```
std::string XMP((char*) buf.pData_+31,size-31);  // or something like that
if ( XMP.find("<iptc:") > 0 ){
    XMP.replace(   "<x:xmpmeta xmlns:x=\"adobe:ns:meta/\">"
               ,   "<x:xmpmeta xmlns:x=\"adobe:ns:meta/\" xmlns:iptc=\"clanmills:ns:iptc\">" );
    memcpy(buf.pData_,(char*)XMP,XMP.size());   // or something like that
}
```

I don't want to fire up Xcode to work on this.  I've hardly written any code for about 3 years.  I'm old and rusty.  However, I'm confident that you'll tame the beast.

**Reply by ribtoks** [reply]
Thank you, Robin!

**Reply by clanmills** [reply]
Thank you, @ribtoks, for bringing up this subject.  It's raining in England, and thinking about this puzzle has been fun.

The answer to the question: _Is there a way to overcome this?_ is **NO**.  It is possible to delete the XMP data.  It's also possible to extract the XMP data, externally repair the XMP, and replace the XMP data.

**Comment by kevinbackhouse** [comment]
You can try commenting out this line of code:

https://github.com/Exiv2/exiv2/blob/bf95347db6e953984e18fa0bf1594827a5005aa7/src/xmp.cpp#L685

It's the source of the error message that you're getting. It's only there as a security mitigation: the Adobe XMP library that we use is very buggy, so the purpose of that line of code is to do a pre-check of the XML to make sure it's valid before we call the Adobe library. If you trust the file that you're running exiv2 on, then you can safely skip the pre-check.

**Reply by ribtoks** [reply]
So you mean that Adobe XMP library will "eat" my invalid XML, just if the exiv2 will allow it?

**Reply by kevinbackhouse** [reply]
Not sure, but I think it's worth a try. Adobe XMP _definitely_ isn't very careful about validating the XML, so it might accept it without crashing.

**Reply by clanmills** [reply]
Gosh, thanks, @kevinbackhouse Kevin. That's valuable insight. I didn't know what you explained.

**Reply by ribtoks** [reply]
It was a good attempt, but still got an error from XMP toolkit

```
Error: XMP Toolkit error 201: XML parsing failure
Warning: Failed to decode XMP metadata.
```

Anyway, thank you for the idea!

**Reply by kevinbackhouse** [reply]
Yes, unfortunately that error message is from the Adobe code:

https://github.com/Exiv2/exiv2/blob/bf95347db6e953984e18fa0bf1594827a5005aa7/xmpsdk/src/ExpatAdapter.cpp#L143

So I think @clanmills suggestion of trying to edit the file to fix the XML is the only option.
